parts/linux/cloud-init/artifacts/cse_install.sh

```diff
@@ -102,12 +103,16 @@ downloadContainerdWasmShims() {
         if [ ! -f "$containerd_wasm_filepath/containerd-shim-spin-${shim_version}" ] || [ ! -f "$containerd_wasm_filepath/containerd-shim-slight-${shim_version}" ]; then
             retrycmd_if_failure 30 5 60 curl -fSLv -o "$containerd_wasm_filepath/containerd-shim-spin-${binary_version}-v1" "$containerd_wasm_url/containerd-shim-spin-v1" 2>&1 | tee $CURL_OUTPUT >/dev/null | grep -E "^(curl:.*)|([eE]rr.*)$" && (cat $CURL_OUTPUT && exit $ERR_KRUSTLET_DOWNLOAD_TIMEOUT)
             retrycmd_if_failure 30 5 60 curl -fSLv -o "$containerd_wasm_filepath/containerd-shim-slight-${binary_version}-v1" "$containerd_wasm_url/containerd-shim-slight-v1" 2>&1 | tee $CURL_OUTPUT >/dev/null | grep -E "^(curl:.*)|([eE]rr.*)$" && (cat $CURL_OUTPUT && exit $ERR_KRUSTLET_DOWNLOAD_TIMEOUT)
-            retrycmd_if_failure 30 5 60 curl -fSLv -o "$containerd_wasm_filepath/containerd-shim-wws-${binary_version}-v1" "$containerd_wasm_url/containerd-shim-wws-v1" 2>&1 | tee $CURL_OUTPUT >/dev/null | grep -E "^(curl:.*)|([eE]rr.*)$" && (cat $CURL_OUTPUT && exit $ERR_KRUSTLET_DOWNLOAD_TIMEOUT)
+            retrycmd_if_failure 30 5 60 curl -fSLv -o "$containerd_wasm_filepath/containerd-shim-wws-${binary_version}-v1" "$containerd_wasm_url/containerd-shim-wws-v1" 2>&1 | tee $CURL_OUTPUT >/dev/null | grep -E "^(curl:.*)|([eE]rr.*)$" && (cat $CURL_OUTPUT && exit $ERR_KRUSTLET_DOWNLOAD_TIMEOUT) &
+            pids+=($!)
             chmod 755 "$containerd_wasm_filepath/containerd-shim-spin-${binary_version}-v1"
             chmod 755 "$containerd_wasm_filepath/containerd-shim-slight-${binary_version}-v1"
-            chmod 755 "$containerd_wasm_filepath/containerd-shim-wws-${binary_version}-v1"
         fi
     done
+    wait ${pids[@]}
+    for shim_version in $CONTAINERD_WASM_VERSIONS; do
+        chmod 755 "$containerd_wasm_filepath/containerd-shim-wws-${binary_version}-v1"
```

Comment:
@lilypan26 Hey! Thank you for the tip on binary_version. I initially took it out because I read the variable would still be in scope, but didn't take into account that its value needs to change as the for loop iterates so it can set permissions on each version.

For the permissions part, the first two downloads (spin / slight) are running sequentially and complete in about a second, so they will definitely be downloaded by the time the program reaches the code that sets permissions for their directories. For the last permission, I had to ensure that the 3 background processes (downloading shim_wws x 3) were complete before reaching that code, otherwise it could attempt to set permissions on a directory that doesn't exist on the VHD yet. I currently only have wws parallelized because downloading the first two and setting their permissions only takes about a second for each binary version.

If you think it makes more sense to parallelize everything for speed or stylistic reasons, please let me know! And thanks again.

Reply:
ah i see, then yea for safety we should only access the directory after we are sure it is created. 

Although for spin and slight, even if they are much faster than wws i think it wouldn't hurt to parallelize those are well and then move all of the chmods to the for loop afterwards. This might be more consistent and readable, but up to you. If not, i would at least add a comment where you are downloading wws and explain that you are parallelizing wws download and only changing the permissions after you are sure the file is there.

Reply:
@lilypan26 I implemented your suggestions, thanks!

That being said, I noticed a flakey little error that has been happening for some time now (before my changes) regarding setting permissions. Im looking into it then will submit another PR.